src/gluonnlp/estimator/machine_translation_event_handler.py

```diff
@@ -0,0 +1,383 @@
+        self.step_num = 0
+        self.avg_start = avg_start
+
+    def _update_avg_param(self, estimator):
+        if estimator.avg_param is None:
+            estimator.avg_param = {k:v.data(estimator.context[0]).copy() for k, v in
+                                   estimator.net.collect_params().items()}
+        if self.step_num > self.avg_start:
+            params = estimator.net.collect_params()
+            alpha = 1. / max(1, self.step_num - self.avg_start)
+            for key, val in estimator.avg_param.items():
+                val[:] += alpha * (params[key].data(estimator.context[0]) - val)
+
+    def epoch_begin(self, estimator, *args, **kwargs):
+        self.batch_id = 0
+                
+    def batch_end(self, estimator, *args, **kwargs):
+        if self.batch_id % self.grad_interval == 0:
+            self.step_num += 1
+        if self.batch_id % self.grad_interval == self.grad_interval - 1:
+            self._update_avg_param(estimator)
+        self.batch_id += 1
+
+    def epoch_end(self, estimator, *args, **kwargs):
+        self._update_avg_param(estimator)
+
+class MTGNMTLearningRateHandler(EpochEnd):
+    def __init__(self, epochs, lr_update_factor):
+        self.epoch_id = 0
+        self.epochs = epochs
+        self.lr_update_factor = lr_update_factor
+
+    def epoch_end(self, estimator, *args, **kwargs):
+        if self.epoch_id + 1 >= (self.epochs * 2) // 3:
+            new_lr = estimator.trainer.learning_rate * self.lr_update_factor
+            estimator.trainer.set_learning_rate(new_lr)
+        self.epoch_id += 1
+
+class TransformerLearningRateHandler(EpochBegin, BatchBegin):
+    def __init__(self, lr,
+                 num_units=512,
+                 warmup_steps=4000,
+                 grad_interval=1):
+        self.lr = lr
+        self.num_units = num_units
+        self.warmup_steps = warmup_steps
+        self.grad_interval = grad_interval
+        self.step_num = 0
+
+    def epoch_begin(self, estimator, *args, **kwargs):
+        self.batch_id = 0
+
+    def batch_begin(self, estimator, *args, **kwargs):
+        if self.batch_id % self.grad_interval == 0:
+            self.step_num += 1
+            new_lr = self.lr /  math.sqrt(self.num_units) * \
+                     min(1. / math.sqrt(self.step_num), self.step_num *
+                         self.warmup_steps ** (-1.5))
+            estimator.trainer.set_learning_rate(new_lr)
+        self.batch_id += 1
+
+class MTGNMTGradientUpdateHandler(GradientUpdateHandler):
+    def __init__(self, clip):
+        super(MTGNMTGradientUpdateHandler, self).__init__()
+        self.clip = clip
+
+    def batch_end(self, estimator, *args, **kwargs):
+        grads = [p.grad(estimator.context[0])
+                 for p in estimator.net.collect_params().values()]
+        gnorm = gluon.utils.clip_global_norm(grads, self.clip)
+        estimator.trainer.step(1)
+
+class TransformerGradientAccumulationHandler(GradientUpdateHandler,
```

Comment:
@eric-haibin-lin

Reply:
Why does grad acc handler API require `batch_size`?

Reply:
It needs batch_size to rescale. See here https://github.com/dmlc/gluon-nlp/blob/master/scripts/machine_translation/train_transformer.py#L320